kubernetes/cache/kube_cache.go

```diff
@@ -1464,11 +1464,21 @@ func (c *kubeCache) GetTelemetries(namespace, labelSelector string) ([]*v1alpha1
 	return retTelemetries, nil
 }
 
+func (c *kubeCache) IsK8sGatewayListerInit(namespace string) bool {
+	// Read lock will prevent the cache from being refreshed while we are reading from the lister
+	// but it won't prevent other routines from reading from the lister.
+	defer c.cacheLock.RUnlock()
+	c.cacheLock.RLock()
+	return c.getCacheLister(namespace).k8sgatewayLister != nil
```

Comment:
This is a larger change that you don't need to implement but `getCacheLister` should probably just return an error to handle scenarios like this. More verbose but less risky.
```
func (c *kubeCache) getCacheLister(namespace string) (*cacheLister, error)
```

Reply:
The `getCacheLister` is just one of the calls alongside with `c.clusterCacheLister` and `c.nsCacheLister`

```
		if c.clusterScoped {
			k8sGateways, err = c.clusterCacheLister.k8sgatewayLister.List(selector)
			if err != nil {
				return nil, err
			}
		} else {
			for _, nsCacheLister := range c.nsCacheLister {
				gatewaysNamespaced, err := nsCacheLister.k8sgatewayLister.List(selector)
				if err != nil {
					return nil, err
				}
				k8sGateways = append(k8sGateways, gatewaysNamespaced...)
			}
		}
	} else {
		k8sGateways, err = c.getCacheLister(namespace).k8sgatewayLister.Gateways(namespace).List(selector)
		if err != nil {
			return nil, err
		}
	}
```
I still think we need to have `lister` existence check outside of the `getCacheLister` method, as the method should not check GW API related listers it has, so keeping the `IsK8sGatewayListerInit` but this time withing the upper caller cache lock:
```
defer c.cacheLock.RUnlock()
	c.cacheLock.RLock()
	if !c.IsK8sGatewayListerInit(namespace) {
		return nil, nil
	}
	g, err := c.getCacheLister(namespace).k8sgatewayLister.Gateways(namespace).Get(name)
	if err != nil {
		return nil, err
	}
```

Reply:
You're right the issue here is the `.k8sgatewayLister` part and not the `getCacheLister(namespace)` part.